core/common/src/test/java/alluxio/resource/PooledResourceTest.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.resource;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+import org.junit.Test;
+
+import java.io.IOException;
+import java.lang.ref.PhantomReference;
+import java.lang.ref.Reference;
+import java.lang.ref.ReferenceQueue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+public class PooledResourceTest {
+  @Test
+  public void releaseOnClose() {
+    TestPool pool = new TestPool(5);
+    try (PooledResource<Resource> i = new PooledResource<>(pool.acquire(), pool)) {
+      assertFalse(i.get().isClosed());
+      assertEquals(0, pool.getAvailableCount());
+      assertEquals(1, pool.getLeakCount());
+    }
+    assertEquals(1, pool.getAvailableCount());
+    assertEquals(0, pool.getLeakCount());
+  }
+
+  @Test
+  public void leakDoesNotPreventPoolFromBeingGCed() throws Exception {
+    long gcTimeout = 1000;
+    TestPool pool = new TestPool(5);
+    ReferenceQueue<TestPool> referenceQueue = new ReferenceQueue<>();
+    PhantomReference<TestPool> poolRef = new PhantomReference<>(pool, referenceQueue);
+    PooledResource<Resource> notLeaked = new PooledResource<>(pool.acquire(), pool);
+    PooledResource<Resource> leaked = new PooledResource<>(pool.acquire(), pool);
+    assertFalse(notLeaked.get().isClosed());
+    assertFalse(leaked.get().isClosed());
+    assertEquals(0, pool.getAvailableCount());
```

Comment:
I am sorry I don't understand this number.
The pool size is 5, and 2 resources have been acquired from the pool.
So, the `AvailableCount()` should be 3?

Reply:
5 is the max number that the pool can contain. At first the pool is empty, the resources are only created at the time they are acquired.

Reply:
OK